marketplace-service/src/main/java/com/axonivy/market/service/impl/MetadataServiceImpl.java

```diff
@@ -0,0 +1,323 @@
+    List<MavenArtifactModel> artifactModelsInVersion =
+        artifactVersionCache.getProductArtifactsByVersion().computeIfAbsent(
+            version, k -> new ArrayList<>());
+    if (metadata.isProductArtifact()) {
+      if (artifactModelsInVersion.stream().anyMatch(artifact -> StringUtils.equals(metadata.getName(),
+          artifact.getName()))) {
+        return;
+      }
+      artifactModelsInVersion.add(
+          MavenUtils.buildMavenArtifactModelFromMetadata(version, metadata));
+    } else {
+      artifactVersionCache.getAdditionalArtifactsByVersion().computeIfAbsent(version, k -> new ArrayList<>()).add(
+          MavenUtils.buildMavenArtifactModelFromMetadata(version, metadata));
+    }
+  }
+
+  private void updateMavenArtifactVersionData(List<String> releasedVersions,
+      Set<Metadata> metadataSet, MavenArtifactVersion artifactVersionCache) {
+    for (Metadata metadata : metadataSet) {
+      String metadataContent = MavenUtils.getMetadataContentFromUrl(metadata.getUrl());
+      if (StringUtils.isBlank(metadataContent)) {
+        continue;
+      }
+      MetadataReaderUtils.parseMetadataFromString(metadataContent, metadata);
+      updateMavenArtifactVersionFromMetadata(artifactVersionCache, metadata);
+      updateContentsFromNonMatchVersions(releasedVersions, metadata);
+    }
+  }
+
+  public void syncAllProductsMetadata() {
+    List<Product> products = productRepo.getAllProductsWithIdAndReleaseTagAndArtifact();
+    log.warn("**MetadataService: Start to sync version for {} product(s)", products.size());
+    for (Product product : products) {
+      // Set up cache before sync
+      String productId = product.getId();
+      Set<Metadata> metadataSet = new HashSet<>(metadataRepo.findByProductId(product.getId()));
+      MavenArtifactVersion artifactVersionCache = mavenArtifactVersionRepo.findById(product.getId()).orElse(
+          MavenArtifactVersion.builder().productId(productId).productArtifactsByVersion(
+              new HashMap<>()).additionalArtifactsByVersion(new HashMap<>()).build());
+      MetadataSync syncCache = metadataSyncRepo.findById(product.getId()).orElse(
+          MetadataSync.builder().productId(product.getId()).syncedVersions(new HashSet<>()).build());
+      Set<Artifact> artifactsFromNewTags = new HashSet<>();
+
+      // Find artifacts from unhandled tags
+      List<String> nonSyncedVersionOfTags = VersionUtils.removeSyncedVersionsFromReleasedVersions(
+          product.getReleasedVersions(), syncCache.getSyncedVersions());
+      if (ObjectUtils.isNotEmpty(nonSyncedVersionOfTags)) {
+        artifactsFromNewTags.addAll(getArtifactsFromNonSyncedVersion(product.getId(), nonSyncedVersionOfTags));
+        log.info("**MetadataService: New tags detected: {} in product {}", nonSyncedVersionOfTags.toString(),
+            productId);
+      }
+
+      // Sync versions from maven & update artifacts-version table
+      List<Artifact> additionalArtifactFromMeta = product.getArtifacts();
```

Comment:
pass product.getArtifacts(); directly to  MavenUtils.convertArtifactsToMetadataSet(new HashSet<>( product.getArtifacts()), productId));